chainstate/src/detail/chainstateref.rs

```diff
@@ -452,46 +459,75 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
+                        tx.get_id(),
+                        block.get_id()
+                    )
+                );
+                ensure!(
+                    block_inputs.insert(input.outpoint()),
+                    CheckBlockTransactionsError::DuplicateInputInBlock(block.get_id())
+                );
             }
         }
+        Ok(())
+    }
+
+    fn check_duplicate_txs(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        // check duplicate transactions
+        let mut txs_ids = BTreeSet::new();
+        for tx in block.transactions() {
+            let tx_id = tx.get_id();
+            let already_in_tx_id = txs_ids.get(&tx_id);
+            match already_in_tx_id {
+                Some(_) => {
+                    return Err(CheckBlockTransactionsError::DuplicatedTransactionInBlock(
+                        tx_id,
+                        block.get_id(),
+                    ))
+                }
+                None => txs_ids.insert(tx_id),
+            };
+        }
+        Ok(())
+    }
 
-        {
-            // check duplicate transactions
-            let mut txs_ids = BTreeSet::new();
-            for tx in block.transactions() {
-                let tx_id = tx.get_id();
-                let already_in_tx_id = txs_ids.get(&tx_id);
-                match already_in_tx_id {
-                    Some(_) => {
-                        return Err(CheckBlockTransactionsError::DuplicatedTransactionInBlock(
-                            tx_id,
-                            block.get_id(),
-                        ))
-                    }
-                    None => txs_ids.insert(tx_id),
-                };
-            }
+    fn check_tokens_txs(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        for tx in block.transactions() {
+            // We can't issue a few token types in one tx
+            let issuance_count = get_tokens_issuance_count(tx);
+            ensure!(
+                issuance_count <= TOKEN_MAX_ISSUANCE_ALLOWED,
+                CheckBlockTransactionsError::TokensError(
+                    TokensError::MultipleTokenIssuanceInTransaction(tx.get_id(), block.get_id()),
+                )
+            );
+
+            // Check tokens
+            tx.outputs()
+                .iter()
+                .filter_map(|output| match output.value() {
+                    OutputValue::Coin(_) => None,
+                    OutputValue::Token(token) => Some(token),
+                })
+                .try_for_each(|token| tokens::check_tokens_data(token, tx, block))
+                .map_err(CheckBlockTransactionsError::TokensError)?;
         }
+        Ok(())
+    }
 
+    fn check_transactions(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        self.check_duplicate_inputs(block)?;
+        self.check_duplicate_txs(block)?;
+        self.check_tokens_txs(block)?;
```

Comment:
I think, there is a question about balance of readable, and efficient. In previous iterations of the review, people told me that we have to split functions to a small one with more clear code. However, over here, two of three functions, use `BTreeSet` as temporary mut value for intermediate computations. If we use through the one cycle, we have to expose mut parameters out of functions. That probably a bad idea. Or we should combine all functions into big one, that seems not a better solution too. Probably, @TheQuantumPhysicist can be pointing us to the right solution.

Reply:
You are right that readability is important here. I would just try to move loop out of functions and leave these 3 function as subroutines doing their checks on a single tx. But again this is more of a suggestion.

Reply:
It doesn't really matter if you loop one or 10 times... every loop iteration costs one extra integer comparison... that's literally nothing even for a raspberry pi. It's better to keep things as independent as possible so that they can be individually tested (if we come to want to do it).